File: iOS/MyAIAgent/MyAIAgentTests/TestSupport.swift
//
//  TestSupport.swift
//  MyAIAgentTests
//

import Foundation
import Synchronization
@testable import MyAIAgent

/// Scripted stand-in at the network boundary. Each `stream()` call consumes
/// the next scripted turn and records the request, so tests can assert on
/// the exact wire conversation the engine sent.
nonisolated final class FakeModelProvider: ModelProvider {
    enum ScriptedTurn {
        case events([StreamEvent])
        case failure(any Error)
        /// A stream that never produces or finishes — for cancellation tests.
        case hang
    }

    private struct State {
        var script: [ScriptedTurn]
        var requests: [[WireMessage]] = []
    }

    private let state: Mutex<State>

    init(script: [ScriptedTurn]) {
        state = Mutex(State(script: script))
    }

    var recordedRequests: [[WireMessage]] {
        state.withLock { $0.requests }
    }

    func stream(
        system: String,
        tools: [ToolDefinition],
        messages: [WireMessage]
    ) -> AsyncThrowingStream<StreamEvent, any Error> {
        let turn: ScriptedTurn? = state.withLock { state in
            state.requests.append(messages)
            return state.script.isEmpty ? nil : state.script.removeFirst()
        }
        return AsyncThrowingStream { continuation in
            switch turn {
            case .events(let events):
                for event in events { continuation.yield(event) }
                continuation.finish()
            case .failure(let error):
                continuation.finish(throwing: error)
            case .hang:
                break  // never yields; consumer cancellation terminates it
            case nil:
                continuation.finish(throwing: AgentError.generic)
            }
        }
    }
}

extension FakeModelProvider.ScriptedTurn {
    /// A plain streamed text reply ending the turn.
    static func textTurn(_ text: String) -> Self {
        .events([.textStarted, .textDelta(text), .finished(stopReason: .endTurn)])
    }

    /// A turn where the model calls one tool with the given JSON input.
    static func toolTurn(id: String, name: String, inputJSON: String) -> Self {
        .events([
            .toolUseStarted(id: id, name: name),
            .toolInputDelta(id: id, partialJSON: inputJSON),
            .toolUseFinished(id: id),
            .finished(stopReason: .toolUse),
        ])
    }
}

// MARK: Engine factory

@MainActor
func makeEngine(
    lead: Lead = .sample,
    provider: FakeModelProvider,
    store: CRMStore? = nil
) -> (engine: AgentEngine, store: CRMStore, outbox: AgentOutbox) {
    let store = store ?? CRMStore(leads: [lead])
    let outbox = AgentOutbox(directory: temporaryOutboxDirectory())
    let engine = AgentEngine(lead: lead, provider: provider, store: store, outbox: outbox)
    return (engine, store, outbox)
}

/// Isolated per-test storage so persistence tests never interfere.
nonisolated func temporaryOutboxDirectory() -> URL {
    FileManager.default.temporaryDirectory
        .appendingPathComponent("outbox-tests-\(UUID().uuidString)", isDirectory: true)
}

/// Polls until the condition holds or the deadline passes — the fake
/// provider is synchronous, so runs settle within a few ticks.
@MainActor
func waitUntil(
    timeout: Duration = .seconds(2),
    _ condition: () -> Bool
) async throws {
    let deadline = ContinuousClock.now + timeout
    while !condition() {
        guard ContinuousClock.now < deadline else {
            throw AgentError.generic  // timeout — the awaited state never arrived
        }
        try await Task.sleep(for: .milliseconds(10))
    }
}

File: android/app/src/test/java/com/curiousbeagle/android/myaiagent/TestSupport.kt
package com.curiousbeagle.android.myaiagent

import com.curiousbeagle.android.myaiagent.agent.AgentEngine
import com.curiousbeagle.android.myaiagent.agent.AgentOutbox
import com.curiousbeagle.android.myaiagent.model.CrmStore
import com.curiousbeagle.android.myaiagent.model.Lead
import com.curiousbeagle.android.myaiagent.net.ModelProvider
import com.curiousbeagle.android.myaiagent.net.StreamEvent
import com.curiousbeagle.android.myaiagent.net.ToolDefinition
import com.curiousbeagle.android.myaiagent.net.WireMessage
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.awaitCancellation
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import java.io.File
import kotlin.io.path.createTempDirectory

/**
 * Scripted stand-in at the network boundary, mirroring the iOS
 * `FakeModelProvider`. Each `stream()` call consumes the next scripted turn
 * and records the request, so tests can assert on the exact wire
 * conversation the engine sent. Single-threaded test dispatchers make the
 * plain lists safe.
 */
class FakeModelProvider(script: List<ScriptedTurn>) : ModelProvider {

    sealed interface ScriptedTurn {
        data class Events(val events: List<StreamEvent>) : ScriptedTurn
        data class Failure(val throwable: Throwable) : ScriptedTurn

        /** A stream that never produces or finishes — for cancellation tests. */
        data object Hang : ScriptedTurn

        companion object {
            /** A plain streamed text reply ending the turn. */
            fun textTurn(text: String) = Events(
                listOf(
                    StreamEvent.TextStarted,
                    StreamEvent.TextDelta(text),
                    StreamEvent.Finished(StreamEvent.StopReason.EndTurn),
                )
            )

            /** A turn where the model calls one tool with the given JSON input. */
            fun toolTurn(id: String, name: String, inputJson: String) = Events(
                listOf(
                    StreamEvent.ToolUseStarted(id, name),
                    StreamEvent.ToolInputDelta(id, inputJson),
                    StreamEvent.ToolUseFinished(id),
                    StreamEvent.Finished(StreamEvent.StopReason.ToolUse),
                )
            )
        }
    }

    private val remaining = script.toMutableList()
    val recordedRequests = mutableListOf<List<WireMessage>>()

    override fun stream(
        system: String,
        tools: List<ToolDefinition>,
        messages: List<WireMessage>,
    ): Flow<StreamEvent> = flow {
        recordedRequests += messages
        when (val turn = if (remaining.isEmpty()) null else remaining.removeAt(0)) {
            is ScriptedTurn.Events -> turn.events.forEach { emit(it) }
            is ScriptedTurn.Failure -> throw turn.throwable
            ScriptedTurn.Hang -> awaitCancellation()
            null -> throw IllegalStateException("script exhausted")
        }
    }
}

/** Isolated per-test storage so persistence tests never interfere. */
fun temporaryOutboxDirectory(): File = createTempDirectory("outbox-tests").toFile()

fun makeEngine(
    scope: CoroutineScope,
    provider: FakeModelProvider,
    lead: Lead = Lead.sample,
    store: CrmStore = CrmStore(listOf(lead)),
    outbox: AgentOutbox = AgentOutbox(temporaryOutboxDirectory()),
): Triple<AgentEngine, CrmStore, AgentOutbox> {
    val engine = AgentEngine(lead, provider, store, outbox, scope)
    return Triple(engine, store, outbox)
}
